src/ShopBundle/Service/ProductVariantService.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+namespace ChameleonSystem\ShopBundle\Service;
+
+use ChameleonSystem\ShopBundle\Interfaces\ProductVariantServiceInterface;
+
+class ProductVariantService implements ProductVariantServiceInterface
+{
+    /**
+     * {@inheritDoc}
+     */
+    public function getProductBasedOnSelection(\TdbShopArticle $shopArticle, array $typeSelection): \TdbShopArticle
```

Comment:
The methods job is to select the variant matching the typeSelection - I think we should be able to simplify the method while increasing performance at the same time by accessing the database directly.

I would suggest:
`getProductBasedOnSelection(string $articleId, array $typeSelection): ?string`

return null if the type selection is not sufficient to select a variante
return the variant article id if the type select is sufficient.

Reply:
Hm, all the method needs is already loaded when calling it (esp. the article itself).

Furthermore the parent is always needed to determine the variant set - ok that could be made with the proper db call but that is not as easily readable und would break the separation of layers as this is no data access class.

Also if only an id is returned the calling method would need to load this object itself still.

Reply:
The method makes quite a lot of db calls
`$shopArticle->GetFieldVariantParent();`
`$shopArticle->GetFieldShopVariantSet();`
`$variantSet->GetFieldShopVariantTypeList();`
all calls wrapped in matchUserSelectionToAvailableVariantValues
`$shopArticle->GetFieldShopArticleVariantsList($typeSelection);`

Using a query I should be able to get all variants that match the current selection as follows:

```sql
SELECT shop_article.*, COUNT(shop_article.id) value_match_count
FROM shop_article_shop_variant_type_value_mlt
INNER JOIN shop_article ON shop_article.id = shop_article_shop_variant_type_value_mlt.source_id
WHERE shop_article_shop_variant_type_value_mlt.target_id IN (VALUE LIST)
AND shop_article.variant_parent_id = parentArticleId
GROUP BY shop_article.id
ORDER BY COUNT(shop_article.id) DESC
```

If we only need articles that match all of the values, then we can add a HAVING COUNT(shop_article.id) = NUMBER-OF-VALUES-IN-LIST

In that case we can skip the order by as well.

If we only need the first match, we can add a limit to further speed things up

Reply:
Now using \TShopArticle::GetFieldShopArticleVariantsList() which does something similar to this query.